.circleci/config.yml

```diff
@@ -91,85 +90,230 @@ oracle-db-container: &oracle-db-container
+
+  prepare-db-for-tests:
+    steps:
+      - run:
+          name: Prepare database for tests
+          command: |
+            echo "Running for database: $DB"
+            bundle exec rake ci:db:ready db:create db:test:prepare
+
+  rspec-tests:
+    parameters:
+      extra-deps:
+        description: "Steps that will be executed in case extra dependencies are required, e.g. to connect to db."
+        type: steps
+        default: []
+    steps:
+      - checkout-with-submodules
+      - *attach-to-workspace
+      - *use-example-config-files
+      - steps: << parameters.extra-deps >>
+      - prepare-db-for-tests
+      - run:
+          name: Rspec tests
+          command: |
+            bundle exec rspec --format progress $(circleci tests glob "spec/**/*_spec.rb" | circleci tests split --split-by=timings)
+      - upload-artifacts
+
+  cucumber-tests:
+    parameters:
+      extra-deps:
+        description: "Steps that will be executed in case extra dependencies are required, e.g. to connect to db."
+        type: steps
+        default: []
+    steps:
+      - checkout-with-submodules
+      - *attach-to-workspace
+      - steps: << parameters.extra-deps >>
+      - prepare-db-for-tests
+      - *disable-internet-access
+      - run:
+          name: Run cucumber tests
+          command: |
+            TESTS=$(circleci tests glob "features/**/*.feature" | circleci tests split --split-by=timings)
+            bundle exec cucumber --profile ci ${TESTS}
+      - *enable-internet-access
+      - upload-artifacts
+      - store_artifacts:
+          path: tmp/capybara
+          destination: capybara
+
+  rails-tests:
+    parameters:
+      extra-deps:
+        description: "Steps that will be executed in case extra dependencies are required, e.g. to connect to db."
+        type: steps
+        default: []
+    steps:
+      - checkout-with-submodules
+      - *attach-to-workspace
+      - *use-example-config-files
+      - steps: << parameters.extra-deps >>
+      - prepare-db-for-tests
+      - run:
+          name: Run Rails tests
+          command: |
+            TESTS=$(bundle exec rake "test:files:${CIRCLE_JOB%-oracle}" | circleci tests split --split-by=timings)
+            bundle exec rake test:run TESTS="$TESTS" TESTOPTS=--verbose --verbose --trace
           environment:
-            BUNDLE_RETRY: 3
-            BUNDLE_JOBS: 3
+            LC_ALL: en_US.UTF-8
+      - upload-artifacts
+
+  save-gem-cache:
+    parameters:
+      database:
+        type: string
+    steps:
       - save_cache:
-          key: v1-bundler-gems-oracle-{{ arch }}-{{ .Branch }}-{{ checksum "Gemfile.lock" }}-{{ checksum "gemfiles/prod/Gemfile.lock" }}
-          paths:
-            - ./vendor/bundle
-            - ./.bundle/
-      - persist_to_workspace:
-          root: .
+          key: v1-bundler-gems-<< parameters.database >>-{{ arch }}-{{ .Branch }}-{{ checksum "Gemfile.lock" }}-{{ checksum "gemfiles/prod/Gemfile.lock" }}
```

Comment:
Probably we would only need a parameter for each Gemfile (upstream, rails50, prod)